File: dev/sbt-plugin/src/sbt-test/sbt-plugin/akka-persistence-typed-migration-java/shopping-cart-lagom-persistence/src/main/java/com/example/shoppingcart/impl/ShoppingCartServiceImpl.java
/*
 * Copyright (C) Lightbend Inc. <https://www.lightbend.com>
 */

package com.example.shoppingcart.impl;

import akka.NotUsed;
import com.example.shoppingcart.api.ShoppingCartService;
import com.lightbend.lagom.javadsl.api.ServiceCall;
import com.lightbend.lagom.javadsl.api.transport.BadRequest;
import com.lightbend.lagom.javadsl.api.transport.NotFound;
import com.lightbend.lagom.javadsl.persistence.PersistentEntityRef;
import com.lightbend.lagom.javadsl.persistence.PersistentEntityRegistry;

import javax.inject.Inject;
import java.util.concurrent.CompletionStage;


public class ShoppingCartServiceImpl implements ShoppingCartService {

    private final PersistentEntityRegistry persistentEntityRegistry;

    private final ReportRepository reportRepository;

    // #akka-persistence-register-classic
    @Inject
    public ShoppingCartServiceImpl(PersistentEntityRegistry persistentEntityRegistry,
                                   ReportRepository reportRepository) {
        this.persistentEntityRegistry = persistentEntityRegistry;
        this.reportRepository = reportRepository;
        persistentEntityRegistry.register(ShoppingCartEntity.class);
    }
    // #akka-persistence-register-classic

    //#akka-persistence-reffor-before
    private PersistentEntityRef<ShoppingCartCommand> entityRef(String id) {
        return persistentEntityRegistry.refFor(ShoppingCartEntity.class, id);
    }

    @Override
    public ServiceCall<NotUsed, String> get(String id) {
        return request ->
            entityRef(id)
                .ask(ShoppingCartCommand.Get.INSTANCE)
                .thenApply(cart -> asShoppingCartView(id, cart));
    }
    //#akka-persistence-reffor-before

    @Override
    public ServiceCall<NotUsed, String> getReport(String id) {
        return request ->
                reportRepository.findById(id).thenApply(report -> {
                    if (report != null) {
                        if (report.isCheckedOut()) return "checkedout";
                        else return "open";

                    }
                    else
                        throw new NotFound("Couldn't find a shopping cart report for '" + id + "'");
                });
    }

    @Override
    public ServiceCall<NotUsed, String> updateItem(String id, String productId, int qty) {
        return item ->
                convertErrors(
                        entityRef(id)
                                .ask(new ShoppingCartCommand.UpdateItem(productId, qty))
                                .thenApply(cart -> asShoppingCartView(id, cart))
                );
    }

    @Override
    public ServiceCall<NotUsed, String> checkout(String id) {
        return request ->
                convertErrors(
                        entityRef(id)
                                .ask(ShoppingCartCommand.Checkout.INSTANCE)
                                .thenApply(cart -> asShoppingCartView(id, cart))
                );
    }


    private <T> CompletionStage<T> convertErrors(CompletionStage<T> future) {
        return future.exceptionally(ex -> {
            if (ex instanceof ShoppingCartException) {
                throw new BadRequest(ex.getMessage());
            }
            else {
                throw new BadRequest("Error updating shopping cart");
            }
        });
    }

    private String asShoppingCartView(String id, Summary summary) {
        StringBuilder builder = new StringBuilder();
        // abc:foo=10:checkedout
        builder.append(id);

        summary.items.forEach((key, value) -> builder
            .append(":")
            .append(key)
            .append("=")
            .append(value));


        if (summary.checkedOut) builder.append(":checkedout");
        else builder.append(":open");

        return builder.toString();
    }

}

File: dev/sbt-plugin/src/sbt-test/sbt-plugin/akka-persistence-typed-migration-scala/shopping-cart-lagom-persistence/src/main/scala/com/example/shoppingcart/impl/ShoppingCartServiceImpl.scala
/*
 * Copyright (C) 2016-2020 Lightbend Inc. <https://www.lightbend.com>
 */

package com.example.shoppingcart.impl

import java.time.OffsetDateTime

import akka.{Done, NotUsed}
import com.example.shoppingcart.api.ShoppingCartService
import com.lightbend.lagom.scaladsl.api.ServiceCall
import com.lightbend.lagom.scaladsl.api.transport.{BadRequest, NotFound, TransportException}
import com.lightbend.lagom.scaladsl.persistence.{EventStreamElement, PersistentEntityRegistry}

import scala.concurrent.ExecutionContext

/**
  * Implementation of the `ShoppingCartService`.
  */
class ShoppingCartServiceImpl(persistentEntityRegistry: PersistentEntityRegistry, reportRepository: ShoppingCartReportRepository)(implicit ec: ExecutionContext) extends ShoppingCartService {

  //#akka-persistence-reffor-before
  /**
    * Looks up the shopping cart entity for the given ID.
    */
  private def entityRef(id: String) =
    persistentEntityRegistry.refFor[ShoppingCartEntity](id)

  override def get(id: String): ServiceCall[NotUsed, String] = ServiceCall { _ =>
    entityRef(id)
      .ask(Get)
      .map(cart => asShoppingCartView(id, cart))
  }
  //#akka-persistence-reffor-before

  override def updateItem(id: String, productId: String, qty: Int): ServiceCall[NotUsed, String] = ServiceCall { update =>
    entityRef(id)
      .ask(UpdateItem(productId, qty))
      .map(cart => asShoppingCartView(id, cart))
      .recover {
        case ShoppingCartException(message) => throw BadRequest(message)
      }
  }

  override def checkout(id: String): ServiceCall[NotUsed, String] = ServiceCall { _ =>
    entityRef(id)
      .ask(Checkout)
      .map(cart => asShoppingCartView(id, cart))
      .recover {
        case ShoppingCartException(message) => throw BadRequest(message)
      }
  }

  private def asShoppingCartView(id: String, cart: Summary): String = {
    val items = cart.items.map {case (k, v) => s"$k=$v"}.mkString(":")
    val status = if (cart.checkedOut) "checkedout" else "open"
    s"$id:$items:$status"
  }

  override def getReport(cartId: String): ServiceCall[NotUsed, String] = ServiceCall { _ =>
    reportRepository.findById(cartId).map {
      case Some(cart) =>
        if (cart.checkedOut) "checkedout"
        else "open"
      case None => throw NotFound(s"Couldn't find a shopping cart report for '$cartId'")
    }
  }
}
